### PR TITLE
[backend] upload docker image into registry

### PR DESCRIPTION
**Pull request by M0ses**

### CODE REVIEW

File: src/backend/bs_publish [hunk 1]
```diff
@@ -2440,3 +2445,92 @@ $conf->{'getflavor'} = $BSConfig::publish_getflavor if $BSConfig::publish_getfla
...
+  my $repository_base = $dr->{repository_base} || '/';
+  $projid =~ s/:/$delimiter/g;
+  my $tag = 'latest';
+  my $up = $dr->{uploader} || 'skopeo';
+  if ($up eq 'skopeo') {
+    my $dst = 'docker://'.$dr->{host}.':'.$dr->{port}.lc($repository_base.$projid."/$repoid/$arch/$name:$tag");
+
+    my $src = "$r/$file";
+    my $tempfile = decompress_container($src);
+    my @cmd = ("skopeo", "copy", "--dest-creds", "$dr->{user}:$dr->{password}", "docker-archive:$tempfile", $dst);
```
Comment (adrianschroeter) [inline src/backend/bs_publish]: Add a FIXME comment that we know the password usage here is a security problem

File: src/backend/bs_publish [hunk 2]
```diff
@@ -2440,3 +2445,92 @@ $conf->{'getflavor'} = $BSConfig::publish_getflavor if $BSConfig::publish_getfla
...
+    BSUtil::printlog("Uploading: '@cmd'");
+    die "Error while uploading\n" if qsystem(@cmd) ) {
+
+    unlink($tempfile);
```
Comment (adrianschroeter) [inline src/backend/bs_publish]: syntax problem, please do also the unlink always independend of success or failure.

File: src/backend/bs_publish [hunk 3]
```diff
@@ -2440,3 +2445,92 @@ $conf->{'getflavor'} = $BSConfig::publish_getflavor if $BSConfig::publish_getfla
...
+  my $ext = $in;
+  my $all_ext = (join('|',keys(%$ext2decomp)));
+  $ext =~ s/.*($all_ext)$/$1/;
+
+  # copy file with cat if no compresison
+  my $decomp = $ext2decomp->{$ext} || 'cat';
+
+  my ($fh, $tempfile) = tempfile();
+
+  system("$decomp $in > $tempfile");
```
Comment (adrianschroeter) [inline src/backend/bs_publish]: potential security problem when the filename can be crafted.

**Comment by M0ses** [comment]
@mschnitzer In a first step, we use 'skopeo' for uploading which handles authentication by itself, so the problem is, how to pass credentials to skopeo without using cli parameters.

**Comment by mschnitzer** [comment]
@M0ses well you could store them in the `config.json` file located in the `.docker` directory of the user. See: https://github.com/projectatomic/skopeo#private-registries-with-authentication

**Comment by adrianschroeter** [comment]
@mschnitzer no this only for the admin. To deal with an own registry instance of the obs instance.

**Comment by mschnitzer** [comment]
Ah ok.

**Pull request merged**